### DISCUSSION TITLE
About the timing of fetching in DeviceTimeAns

### DISCUSSION THREAD

**Discussion by zouOrange** [Lora-net/LoRaMac-node#1388] [Q&A]
When I use the DEVICE_TIME_REQ command, in SRV_MAC_DEVICE_TIME_ANS I read that the returned GPS time is 1669186181 and the converted UTC time is Nov 27, 2032 06:49:23. Obviously this is the wrong value. Why does this happen?Is this time generated by the gateway’s GPS or by the server?

**Comment by mluis1** [comment]
The value `1669186181` looks to be the UTC time. This value corresponds to `1353221399` GPS time.
It looks to be confirmed by the below result obtained by using https://www.andrews.edu/~tzs/timeconv/timeconvert.php

![image](https://user-images.githubusercontent.com/5763295/203524686-ac509f94-5c8b-4785-b888-35ca403a4dfb.png)

The LoRaWAN specifications say that the value of `SRV_MAC_DEVICE_TIME_ANS` must be the GPS time.

![image](https://user-images.githubusercontent.com/5763295/203523386-c6192318-8039-4ce3-ba68-6b1afb895177.png)

The handling of `SRV_MAC_DEVICE_TIME_ANS` is done as follows under this project
https://github.com/Lora-net/LoRaMac-node/blob/23bb8cf67902ede0b9e28f2999c23a0ee1f472c7/src/mac/LoRaMac.c#L2537-L2578

As  you can observe the expected received time is the time since GPS epoch. Which then is converted to UNIX time for the system clock. (Please note that it excludes the leap seconds. Meaning that the UNIX time will be off by a few seconds)

Where have you obtained this value?
Is it extracted from the server logs for downlink?
Is it displayed on the end-device upon reception of the `SRV_MAC_DEVICE_TIME_ANS`?

In the future it would be nice if you could post this kind of questions on the project [Discussions](https://github.com/Lora-net/LoRaMac-node/discussions) tab. It is a better place to engage discussions and then we can agree if it is an issue or not.
If you don't mind I move this issue to the [Discussions](https://github.com/Lora-net/LoRaMac-node/discussions) tab.

**Reply by zouOrange** [reply]
Can you give me some advice? About returning BEACON_STATE_LOST.

**Comment by zouOrange** [comment]
Yes, I get this value in `SRV_MAC_DEVICE_TIME_ANS `in end-device.  
```ruby
sysTime = SysTimeAdd( sysTimeCurrent, SysTimeSub( sysTime, MacCtx.LastTxSysTime ) );
printf("time=%u.%d\r\n", (unsigned int)sysTime.Seconds, 
```

Because I did not appear `LORAMAC_EVENT_INFO_STATUS_OK` in MlmeConfirm when I entered ClassB. Whether it has anything to do with the time gained. So I checked` SRV_MAC_DEVICE_TIME_ANS`.

In `LoRaMacClassBProcessBeacon `function, I capture the `BEACON_STATE_LOST`.
It would be nice if you had any suggestions。

```ruby

static void MlmeConfirm( MlmeConfirm_t *mlmeConfirm )
{
    LmHandlerPackagesNotify( PACKAGE_MLME_CONFIRM, mlmeConfirm );
    switch( mlmeConfirm->MlmeRequest )
   {
     case MLME_BEACON_ACQUISITION:
    {
         if( mlmeConfirm->Status == LORAMAC_EVENT_INFO_STATUS_OK )
         {
             LmHandlerPingSlotReq( 3 );
             printf("BEA_ACQ_OK\r\n");
         }
       else
      {
        // Beacon not acquired
       // Request Device Time again.
        LmHandlerDeviceTimeReq( );
        printf("BEA_ACQ_FAIL\r\n");
      }
  }
   break;
}
```
I certainly don't mind moving to the [Discussions](https://github.com/Lora-net/LoRaMac-node/discussions) tab.
